pkg/plugins/runtime/gateway/route_table_generator.go

```diff
@@ -110,3 +120,108 @@ func (r *RouteTableGenerator) GenerateHost(ctx xds_context.Context, info *Gatewa
 
 	return resources.Get(), nil
 }
+
+// retryRouteConfigurers returns the set of route configurers needed to implement the retry policy (if there is one).
+func retryRouteConfigurers(protocol core_mesh.Protocol, retry *core_mesh.RetryResource) []route.RouteConfigurer {
+	if retry == nil {
+		return nil
+	}
+
+	methodStrings := func(methods []mesh_proto.HttpMethod) []string {
+		var names []string
+		for _, m := range methods {
+			if m != mesh_proto.HttpMethod_NONE {
+				names = append(names, m.String())
+			}
+		}
+		return names
+	}
+
+	grpcConditionStrings := func(conditions []mesh_proto.Retry_Conf_Grpc_RetryOn) []string {
+		var names []string
+		for _, c := range conditions {
+			names = append(names, c.String())
+		}
+		return names
+	}
+
+	configurers := []route.RouteConfigurer{
+		route.RouteActionRetryDefault(protocol),
+	}
+
+	switch protocol {
+	case core_mesh.ProtocolHTTP, core_mesh.ProtocolHTTP2:
+		conf := retry.Spec.GetConf().GetHttp()
+		configurers = append(configurers,
+			route.RouteActionRetryOnStatus(conf.GetRetriableStatusCodes()...),
+			route.RouteActionRetryMethods(methodStrings(conf.GetRetriableMethods())...),
+			route.RouteActionRetryTimeout(conf.GetPerTryTimeout().AsDuration()),
+			route.RouteActionRetryCount(conf.GetNumRetries().GetValue()),
+			route.RouteActionRetryBackoff(
+				conf.GetBackOff().GetBaseInterval().AsDuration(),
+				conf.GetBackOff().GetMaxInterval().AsDuration()),
+		)
+	case core_mesh.ProtocolGRPC:
+		conf := retry.Spec.GetConf().GetGrpc()
+		configurers = append(configurers,
+			route.RouteActionRetryOnConditions(grpcConditionStrings(conf.GetRetryOn())...),
+			route.RouteActionRetryTimeout(conf.GetPerTryTimeout().AsDuration()),
+			route.RouteActionRetryCount(conf.GetNumRetries().GetValue()),
+			route.RouteActionRetryBackoff(
+				conf.GetBackOff().GetBaseInterval().AsDuration(),
+				conf.GetBackOff().GetMaxInterval().AsDuration()),
+		)
+	}
```

Comment:
what's the advantage of creating configurer for every separate field than to create one for retry action based on the Retry policy like we use for regular mesh?

I'm not criticizing this decision, just want to understand the reasons and whether we should follow this pattern for other things.

Reply:
Yeh, definitely worth a discussion. In most cases, the Kuma API is handled at the layer above the configurers, so this is consistent with that approach. IIUC, the layering there is that the generators know about the Kuma API semantics, and the configurers only know about how to instantiate specific Envoy protobuf messages. Retry seems to be an outlier, though I can see why the author took that approach. There wasn't an obvious intermediate data format that would make sense for a single configurer, so I broke it down into smaller fields.